frame/support/procedural/src/construct_runtime/expand/origin.rs

```diff
@@ -0,0 +1,341 @@
+					Err(self)
+				}
+			}
+		}
+		impl From<Option<<#runtime as #system_path::Config>::AccountId>> for Origin {
+			/// Convert to runtime origin with caller being system signed or none and use filter
+			/// `frame-system::Config::BaseCallFilter`.
+			fn from(x: Option<<#runtime as #system_path::Config>::AccountId>) -> Self {
+				<#system_path::Origin<#runtime>>::from(x).into()
+			}
+		}
+
+		#pallet_conversions
+	})
+}
+
+fn expand_origin_caller_variant(
+	runtime: &Ident,
+	path: &PalletPath,
+	index: u8,
+	instance: Option<&Ident>,
+	generics: &Generics,
+) -> TokenStream {
+	let pallet_is_generic = !generics.params.is_empty();
+	let mod_name = &path.mod_name();
+
+	match (instance, pallet_is_generic) {
+		(Some(inst), true) => {
+			let variant = format_ident!("{}_{}", mod_name, inst);
+			quote!(#[codec(index = #index)] #variant(#path::Origin<#runtime, #path::#inst>),)
+		}
+		(Some(inst), false) => {
+			let variant = format_ident!("{}_{}", mod_name, inst);
+			quote!(#[codec(index = #index)] #variant(#path::Origin<#path::#inst>),)
+		}
+		(None, true) => {
+			quote!(#[codec(index = #index)] #mod_name(#path::Origin<#runtime>),)
+		}
+		(None, false) => {
+			quote!(#[codec(index = #index)] #mod_name(#path::Origin),)
+		}
+	}
+}
+
+fn expand_origin_pallet_conversions(
+	scrate: &TokenStream,
+	runtime: &Ident,
+	path: &PalletPath,
+	instance: Option<&Ident>,
+	generics: &Generics,
+) -> TokenStream {
+	let mod_name = path.mod_name();
+	let variant = if let Some(inst) = instance {
+		format_ident!("{}_{}", mod_name, inst)
+	} else {
+		mod_name
+	};
+
+	let pallet_is_generic = !generics.params.is_empty();
+	let pallet_origin = match (instance, pallet_is_generic) {
+		(Some(inst), true) => quote!(#path::Origin<#runtime, #path::#inst>),
+		(Some(inst), false) => quote!(#path::Origin<#path::#inst>),
+		(None, true) => quote!(#path::Origin<#runtime>),
+		(None, false) => quote!(#path::Origin),
+	};
+
+	quote!{
+		impl From<#pallet_origin> for OriginCaller {
+			fn from(x: #pallet_origin) -> Self {
+				OriginCaller::#variant(x)
+			}
+		}
+
+		impl From<#pallet_origin> for Origin {
+			/// Convert to runtime origin using `frame-system::Config::BaseCallFilter`.
+			fn from(x: #pallet_origin) -> Self {
+				let x: OriginCaller = x.into();
+				x.into()
+			}
+		}
+
+		impl Into<#scrate::sp_std::result::Result<#pallet_origin, Origin>> for Origin {
+			/// NOTE: converting to pallet origin loses the origin filter information.
+			fn into(self) -> #scrate::sp_std::result::Result<#pallet_origin, Self> {
+				if let OriginCaller::#variant(l) = self.caller {
+					Ok(l)
+				} else {
+					Err(self)
+				}
+			}
+		}
+
+		impl #scrate::sp_std::convert::TryFrom<OriginCaller> for #pallet_origin {
+			type Error = OriginCaller;
+			fn try_from(
+				x: OriginCaller,
+			) -> #scrate::sp_std::result::Result<#pallet_origin, OriginCaller> {
+				if let OriginCaller::#variant(l) = x {
+					Ok(l)
+				} else {
+					Err(x)
+				}
+			}
+		}
+	}
+}
```

Comment:
This is a bit difficult, since `construct_runtime!` will always generate code that assumes the Call config associated type is specified as the Call enum that it itself generates. More concretely, it's this piece of code in `decl_outer_dispatch!` that is problematic:

```rust
impl $crate::dispatch::Dispatchable for $call_type {
	// ..
	fn dispatch(
		self,
		origin: $origin,
	) -> $crate::dispatch::DispatchResultWithPostInfo {
		if !<Self::Origin as $crate::traits::OriginTrait>::filter_call(&origin, &self) { // <--- this line where it sends &self as the 2nd arg to `filter_call`
			return $crate::sp_std::result::Result::Err($crate::dispatch::DispatchError::BadOrigin.into())
		}

		$crate::traits::UnfilteredDispatchable::dispatch_bypass_filter(self, origin)
	}
}
```

So even if I configure the runtime to use `u32` as the type for `type Call`, the code above would not respect my configuration.

Reply:
ah yes, the tests I am interested in are the tests for the filter stuff, that the BaseCallFilter is correctly applied everywhere except root etc..

So basically just this test:
```rust
	#[test]
	fn test_default_filter() {
		assert_eq!(OriginWithSystem::root().filter_call(&0), true);
		assert_eq!(OriginWithSystem::root().filter_call(&1), true);
		assert_eq!(OriginWithSystem::none().filter_call(&0), true);
		assert_eq!(OriginWithSystem::none().filter_call(&1), false);
		assert_eq!(OriginWithSystem::signed(0).filter_call(&0), true);
		assert_eq!(OriginWithSystem::signed(0).filter_call(&1), false);
		assert_eq!(OriginWithSystem::from(Some(0)).filter_call(&0), true);
		assert_eq!(OriginWithSystem::from(Some(0)).filter_call(&1), false);
		assert_eq!(OriginWithSystem::from(None).filter_call(&0), true);
		assert_eq!(OriginWithSystem::from(None).filter_call(&1), false);
		assert_eq!(OriginWithSystem::from(origin_without_generic::Origin).filter_call(&0), true);
		assert_eq!(OriginWithSystem::from(origin_without_generic::Origin).filter_call(&1), false);

		let mut origin = OriginWithSystem::from(Some(0));

		origin.add_filter(|c| *c % 2 == 1);
		assert_eq!(origin.filter_call(&0), false);
		assert_eq!(origin.filter_call(&1), false);

		origin.set_caller_from(OriginWithSystem::root());
		assert!(matches!(origin.caller, OriginWithSystemCaller::system(frame_system::RawOrigin::Root)));
		assert_eq!(origin.filter_call(&0), false);
		assert_eq!(origin.filter_call(&1), false);

		origin.reset_filter();
		assert_eq!(origin.filter_call(&0), true);
		assert_eq!(origin.filter_call(&1), false);
	}
```

Reply:
I can import this test and use FRAMEv1 macros to run it, but it wouldn't exercise `expand_outer_origin` for the test due to the problem I mentioned in my previous comment.

Reply:
This is actually an argument to not generate any code when `Call` is not included in the runtime, since it's `decl_outer_dispatch` that generated unnecessary code that led to the compilation error.

Reply:
maybe you can put the where clause on the trait implementation itself, not the function.

But I don't understand why it is not possible to use FRAMEv2.

I want a test which tests that the code generated for the origin and filter is correct. this PR duplicate the code generation for the origin and filter from frame/support/src/origin.rs to construct_runtime. But no test ensure the implementation in construct_runtime is working.

So something like this:
```diff
diff --git a/frame/support/test/tests/pallet.rs b/frame/support/test/tests/pallet.rs
index 7478da189d..7400d01d0e 100644
--- a/frame/support/test/tests/pallet.rs
+++ b/frame/support/test/tests/pallet.rs
@@ -1058,3 +1058,36 @@ fn test_storage_info() {
                ],
        );
 }
+
+#[test]
+fn test_default_filter() {
+       use frame_support::traits::OriginTrait;
+
+       assert_eq!(Origin::root().filter_call(&0), true);
+       assert_eq!(Origin::root().filter_call(&1), true);
+       assert_eq!(Origin::none().filter_call(&0), true);
+       assert_eq!(Origin::none().filter_call(&1), false);
+       assert_eq!(Origin::signed(0).filter_call(&0), true);
+       assert_eq!(Origin::signed(0).filter_call(&1), false);
+       assert_eq!(Origin::from(Some(0)).filter_call(&0), true);
+       assert_eq!(Origin::from(Some(0)).filter_call(&1), false);
+       assert_eq!(Origin::from(None).filter_call(&0), true);
+       assert_eq!(Origin::from(None).filter_call(&1), false);
+       assert_eq!(Origin::from(pallet::Origin::default()).filter_call(&0), true);
+       assert_eq!(Origin::from(pallet::Origin::default()).filter_call(&1), false);
+
+       let mut origin = Origin::from(Some(0));
+
+       origin.add_filter(|c| *c % 2 == 1);
+       assert_eq!(origin.filter_call(&0), false);
+       assert_eq!(origin.filter_call(&1), false);
+
+       origin.set_caller_from(Origin::root());
+       assert!(matches!(origin.caller, Origin::system(frame_system::RawOrigin::Root)));
+       assert_eq!(origin.filter_call(&0), false);
+       assert_eq!(origin.filter_call(&1), false);
+
+       origin.reset_filter();
+       assert_eq!(origin.filter_call(&0), true);
+       assert_eq!(origin.filter_call(&1), false);
+}
```
Something which test that the filter is correctly applied and base call filter is correctly applied, and that root bypass the filter

Reply:
`construct_runtime!` generates a `Dispatchable` impl that doesn't respect the type specified in  `Config::Call`. Going back to my previous comment, it's this line here generated by `decl_outer_dispatch` that is problematic:

 ```rust
if !<Self::Origin as $crate::traits::OriginTrait>::filter_call(&origin, &self) {
```

The call to `filter_call` here assumes that the 2nd argument is of type `enum Call`, but in reality it really is specified as `u32` in `Config::Call`, which will lead to a compilation error.

Reply:
Ok, I didn't want to say that the test is exactly this. The test should ensure the implementation of filter_call is accurate, there is a way to execute filter_call in test no ?

Something like this:
```diff
diff --git a/frame/support/test/tests/pallet.rs b/frame/support/test/tests/pallet.rs
index 7478da189d..5db3eebda3 100644
--- a/frame/support/test/tests/pallet.rs
+++ b/frame/support/test/tests/pallet.rs
@@ -1058,3 +1058,38 @@ fn test_storage_info() {
                ],
        );
 }
+
+#[test]
+fn test_default_filter() {
+       use frame_support::traits::OriginTrait;
+       let accepted_call = todo!();
+       let refused_call = todo!();
+
+       assert_eq!(Origin::root().filter_call(&accepted_call), true);
+       assert_eq!(Origin::root().filter_call(&refused_call), true);
+       assert_eq!(Origin::none().filter_call(&accepted_call), true);
+       assert_eq!(Origin::none().filter_call(&refused_call), false);
+       assert_eq!(Origin::signed(0).filter_call(&accepted_call), true);
+       assert_eq!(Origin::signed(0).filter_call(&refused_call), false);
+       assert_eq!(Origin::from(Some(0)).filter_call(&accepted_call), true);
+       assert_eq!(Origin::from(Some(0)).filter_call(&refused_call), false);
+       assert_eq!(Origin::from(None).filter_call(&accepted_call), true);
+       assert_eq!(Origin::from(None).filter_call(&refused_call), false);
+       assert_eq!(Origin::from(pallet::Origin::default()).filter_call(&accepted_call), true);
+       assert_eq!(Origin::from(pallet::Origin::default()).filter_call(&refused_call), false);
+
+       let mut origin = Origin::from(Some(0));
+
+       origin.add_filter(|c| todo!());
+       assert_eq!(origin.filter_call(&accepted_call), false);
+       assert_eq!(origin.filter_call(&refused_call), false);
+
+       origin.set_caller_from(Origin::root());
+       assert!(matches!(origin.caller, Origin::system(frame_system::RawOrigin::Root)));
+       assert_eq!(origin.filter_call(&accepted_call), false);
+       assert_eq!(origin.filter_call(&refused_call), false);
+
+       origin.reset_filter();
+       assert_eq!(origin.filter_call(&accepted_call), true);
+       assert_eq!(origin.filter_call(&refused_call), false);
+}
```

Reply:
Yeah, that would work, I considered that as well, but it would've been really great if we were able to just have `Call` be `u32` and be done with it.

The other issue is, `decl_outer_dispatch` really should be respecting the type given to it via `system::Config::Call`, but that's kind of out of scope for this PR. Not sure if I should file an issue about it, since I think #8084 would solve this.